File: Java/src/bin/easy_73.java
// Daily Coding Problem: 73
// Difficulty: Easy
// Asked by: Google
// Authored by: TenthEdict

class Easy73 {

    public static void main(String[] args) {
        Node node3 = new Node(3, null);
        Node node2 = new Node(2, node3);
        Node node1 = new Node(1, node2);

    printList(node1);
    System.out.println("---");
    printList(reverse(node1));
    }

    static class Node {
        int val;
        Node next;

        Node(int val, Node next) {
            this.val = val;
            this.next = next;
        }
    }

    static Node reverse(Node head) {
        Node prev = null;
        Node curr = head;

        while (curr != null) {
            Node next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
}

    static void printList(Node head) {
        while (head != null) {
            System.out.println(head.val);
            head = head.next;
        }
    }
}

File: C/src/bin/easy_73.c
// Daily Coding Problem: 73
// Difficulty: Easy
// Asked by: Google
// Authored by: TenthEdict

#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    int val;
    struct Node* next;
} Node;

Node* reverse(Node* head)
{
    Node* prev = NULL;
    Node* curr = head;

    while (curr)
    {
        Node* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

void print_list(Node* head)
{
    while (head)
    {
        printf("%d\n", head->val);

        head = head->next;
    }
}

int main()
{
    Node* ptr3 = malloc(sizeof(Node));
    ptr3->val = 3;
    ptr3->next = NULL;

    Node* ptr2 = malloc(sizeof(Node));
    ptr2->val = 2;
    ptr2->next = ptr3;

    Node* ptr1 = malloc(sizeof(Node));
    ptr1->val = 1;
    ptr1->next = ptr2;

    print_list(ptr1);
    Node* reversed = reverse(ptr1);
    print_list(reversed);

    return 0;
}
